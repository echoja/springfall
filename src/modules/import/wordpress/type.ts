export type WordpressJSON = {
  rss?: {
    channel?: {
      item?: {
        title?: string;
        "wp:post_id"?: string;
        "wp:status"?: "publish" | "inherit";
        // "Tue, 26 Jan 2021 12:44:15 +0000",
        pubDate?: string;
        "content:encoded"?: string;
        "wp:post_type"?: "post" | "page" | "attachment";
      }[];
    };
  };
};

export type WordpressItem = {
  title: string;
  content: string;
  pubDate: string;
  post_id: string;
};
