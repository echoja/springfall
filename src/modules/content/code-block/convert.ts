import type { RefractorElement, RefractorRoot, Text } from "refractor";
import { refractor } from "refractor";
import shLang from "refractor/lang/bash";
import cppLang from "refractor/lang/cpp";
import cssLang from "refractor/lang/css";
import htmlLang from "refractor/lang/markup";
import pythonLang from "refractor/lang/python";
import rustLang from "refractor/lang/rust";
import tsxLang from "refractor/lang/tsx";

import type {
  ICodeBlock,
  ICodeElement,
  ICodeLine,
  IText,
  Language,
} from "../types";
import { deepclone } from "@common/util";

import codeNodeToString from "./code-node-to-string";

refractor.register(tsxLang);
refractor.register(rustLang);
refractor.register(pythonLang);
refractor.register(htmlLang);
refractor.register(shLang);
refractor.register(cppLang);
refractor.register(cssLang);

export function getText(node: Text):
  | {
      left: null | Text;
      reachedNewline: true;
      right: null | Text;
    }
  | { reachedNewline: false; node: null | Text } {
  const resultNode = deepclone(node);
  const newlineIndex = node.value.indexOf("\n");
  if (newlineIndex === -1) {
    return {
      node: resultNode,
      reachedNewline: false,
    };
  }
  const beforeNewline = resultNode.value.substring(0, newlineIndex);
  const afterNewline = resultNode.value.substring(newlineIndex + 1);
  resultNode.value = beforeNewline;
  const restNode = deepclone(node);
  restNode.value = afterNewline;
  return {
    left: beforeNewline === "" ? null : resultNode,
    reachedNewline: true,
    right: afterNewline === "" ? null : restNode,
  };
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function getElement(node: RefractorElement):
  | {
      left: RefractorElement | null;
      reachedNewline: true;
      right: RefractorElement | null;
    }
  | {
      node: RefractorElement | null;
      reachedNewline: false;
    } {
  const result: {
    left: RefractorElement | null;
    reachedNewline: true;
    right: RefractorElement | null;
  } = {
    left: null,
    reachedNewline: true,
    right: null,
  };

  result.left?.children.slice();

  const leftChildren: (RefractorElement | Text)[] = [];

  const noNewline = node.children.every((child, index) => {
    const childResult =
      child.type === "element" ? getElement(child) : getText(child);

    if (childResult.reachedNewline && childResult.left) {
      leftChildren.push(childResult.left);
    } else if (!childResult.reachedNewline && childResult.node) {
      leftChildren.push(childResult.node);
    }

    if (!childResult.reachedNewline) {
      return true;
    }

    const rightNode = deepclone(node);
    const leftNode = deepclone(node);
    const rightChildren = node.children.slice(index + 1);
    if (childResult.right) {
      rightChildren.unshift(childResult.right);
    }
    leftNode.children = leftChildren;
    rightNode.children = rightChildren;

    result.reachedNewline = true;
    result.left = leftNode.children.length !== 0 ? leftNode : null;
    result.right = rightNode.children.length !== 0 ? rightNode : null;

    return false;
  });

  if (noNewline) {
    return {
      reachedNewline: false,
      node: deepclone(node),
    };
  }

  return result;
}

export function convertElement(
  node: RefractorElement | Text
): ICodeElement | IText {
  if (node.type === "element") {
    return {
      type: "CODE_ELEMENT",
      tagName: "span",
      children: node.children.map(convertElement),
      properties: deepclone(node.properties),
    };
  }
  return {
    type: "TEXT",
    text: node.value,
  };
}

export function getLine(root: RefractorRoot): {
  children: (RefractorElement | Text)[];
  reachedNewline: boolean;
  restRoot: RefractorRoot | null;
} {
  const result: ReturnType<typeof getLine> = {
    children: [],
    reachedNewline: false,
    restRoot: deepclone(root),
  };

  const noNewline = root.children.every((child, index) => {
    const childResult =
      child.type === "element" ? getElement(child) : getText(child);

    if (childResult.reachedNewline && childResult.left) {
      result.children.push(childResult.left);
    } else if (!childResult.reachedNewline && childResult.node) {
      result.children.push(childResult.node);
    }

    if (!childResult.reachedNewline) {
      return true;
    }

    const newChildren = root.children.slice(index + 1);
    if (childResult.right) {
      newChildren.unshift(childResult.right);
    }
    if (result.restRoot) {
      result.restRoot.children = newChildren;
    }
    result.reachedNewline = true;

    return false;
  });

  if (noNewline) {
    return {
      children: deepclone(root.children),
      reachedNewline: false,
      restRoot: null,
    };
  }

  return result;
}

export function getLines(root: RefractorRoot) {
  const result: (RefractorElement | Text)[][] = [];

  let line: ReturnType<typeof getLine> = {
    children: [],
    reachedNewline: true,
    restRoot: root,
  };
  while (line.restRoot) {
    line = getLine(line.restRoot);
    result.push(line.children);
  }

  return result;
}

export function unifyChild(element: ICodeElement | IText): ICodeElement {
  if (element.type === "TEXT") {
    return {
      type: "CODE_ELEMENT",
      children: [deepclone(element)],
      tagName: "span",
      properties: {
        className: ["token text-wrapper"],
      },
    };
  }

  const isAllText = element.children.every((child) => child.type === "TEXT");
  if (isAllText) {
    return element;
  }
  return {
    ...element,
    children: element.children.map(unifyChild),
  };
}

export function unifyLine(line: ICodeLine): ICodeLine {
  return {
    ...line,
    children: line.children.map(unifyChild),
  };
}

export function unifyCodeBlock(codeBlock: ICodeBlock): ICodeBlock {
  return {
    ...codeBlock,
    children: codeBlock.children.map(unifyLine),
  };
}

export function convertRefactorToCodeBlock(
  root: RefractorRoot,
  lang: Language
): ICodeBlock {
  const lines = getLines(root);
  const codeLines: ICodeLine[] = lines.map((line) => {
    return {
      type: "CODE_LINE",
      children: line.map(convertElement),
    };
  });

  const codeBlock: ICodeBlock = {
    type: "CODE_BLOCK",
    children: codeLines,
    lang,
    showCopy: true,
    showLines: false,
  };

  return unifyCodeBlock(codeBlock);
}

export function createCodeBlock(value: string, language: Language) {
  let formatted: RefractorRoot;
  if (language === "plaintext") {
    formatted = {
      type: "root",
      children: [
        {
          type: "element",
          tagName: "span",
          properties: {
            className: ["token", "text-wrapper"],
          },
          children: [
            {
              type: "text",
              value,
            },
          ],
        },
      ],
    };
  } else {
    formatted = refractor.highlight(value, language);
  }
  return convertRefactorToCodeBlock(formatted, language);
}

export function convert(source: ICodeBlock, language: Language) {
  const content = codeNodeToString(source);
  return createCodeBlock(content, language);
}
