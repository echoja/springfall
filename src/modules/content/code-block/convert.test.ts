import type { RefractorElement, RefractorRoot, Text } from "refractor";
import { describe } from "vitest";

import type { ICodeLine } from "../types";

import { getElement, getLines, getText, unifyLine } from "./convert";

describe("convert", () => {
  describe("getText", () => {
    test("\\n이 있을 때 잘 분리되어야 함", () => {
      const node: Text = {
        type: "text",
        value: "abc\ndef",
      };
      const result = getText(node);
      expect(result).toMatchSnapshot();
    });

    test("\\n이 앞에 있을 때 잘 분리되어야 함", () => {
      const node: Text = {
        type: "text",
        value: "\ndef",
      };
      const result = getText(node);
      expect(result).toMatchSnapshot();
    });

    test("\\n이 뒤에 있을 때 잘 분리되어야 함", () => {
      const node: Text = {
        type: "text",
        value: "abcdef\n",
      };
      const result = getText(node);
      expect(result).toMatchSnapshot();
    });

    test("\\n이 없을 때는 그대로 리턴되어야 함", () => {
      const node: Text = {
        type: "text",
        value: "abcdef",
      };
      const result = getText(node);
      expect(result).toMatchSnapshot();
    });
  });

  describe("getElement", () => {
    test("children 이 없을 떄", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span",
        children: [],
      };
      const result = getElement(node);

      expect(result).toMatchSnapshot();
    });

    test("children 내에 \n이 없을 때", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span",
        children: [
          {
            type: "text",
            value: "abc",
          },
          {
            type: "text",
            value: "dev",
          },
        ],
      };
      const result = getElement(node);
      expect(result).toMatchSnapshot();
    });

    test("element 안에 \\n 이 포함된 Text 가 있을 떄 제대로 동작해야 함", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span",
        children: [
          {
            type: "text",
            value: "abc",
          },
          {
            type: "text",
            value: "def\nghi",
          },
          {
            type: "text",
            value: "jkl",
          },
        ],
      };
      const result = getElement(node);
      expect(result).toMatchSnapshot();
    });

    test("중첩된 경우", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span1",
        children: [
          {
            type: "element",
            tagName: "span2",
            children: [
              {
                type: "element",
                tagName: "span3",
                children: [
                  {
                    type: "text",
                    value: "abc\ndef",
                  },
                ],
              },
            ],
          },
        ],
      };
      const result = getElement(node);
      expect(result).toMatchSnapshot();
    });

    test("중첩되었을 떄 마지막에 \\n이 있을 때", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span1",
        children: [
          {
            type: "element",
            tagName: "span2",
            children: [
              {
                type: "element",
                tagName: "span3",
                children: [
                  {
                    type: "text",
                    value: "abcdef\n",
                  },
                ],
              },
            ],
          },
        ],
      };
      const result = getElement(node);
      expect(result).toMatchSnapshot();
    });

    test("\\n이 여러 개 있을 때", () => {
      const node: RefractorElement = {
        type: "element",
        tagName: "span1",
        children: [
          {
            type: "text",
            value: "abc\ndef",
          },
          {
            type: "element",
            tagName: "span2",
            children: [
              {
                type: "text",
                value: "ghi\njkl",
              },
            ],
          },
        ],
      };
      const result = getElement(node);
      expect(result).toMatchSnapshot();
    });
  });

  describe("getLines", () => {
    test("제대로 동작해야 함", () => {
      const root: RefractorRoot = {
        type: "root",
        children: [
          {
            type: "element",
            tagName: "span",
            properties: {
              className: ["token", "keyword"],
            },
            children: [
              {
                type: "text",
                value: "let",
              },
            ],
          },
          {
            type: "text",
            value: " a ",
          },
          {
            type: "element",
            tagName: "span",
            properties: {
              className: ["token", "operator"],
            },
            children: [
              {
                type: "text",
                value: "=",
              },
            ],
          },
          {
            type: "text",
            value: " ",
          },
          {
            type: "element",
            tagName: "span",
            properties: {
              className: ["token", "template-string"],
            },
            children: [
              {
                type: "element",
                tagName: "span",
                properties: {
                  className: ["token", "template-punctuation", "string"],
                },
                children: [
                  {
                    type: "text",
                    value: "`",
                  },
                ],
              },
              {
                type: "element",
                tagName: "span",
                properties: {
                  className: ["token", "string"],
                },
                children: [
                  {
                    type: "text",
                    value: "\n   sdkfj\n",
                  },
                ],
              },
              {
                type: "element",
                tagName: "span",
                properties: {
                  className: ["token", "template-punctuation", "string"],
                },
                children: [
                  {
                    type: "text",
                    value: "`",
                  },
                ],
              },
            ],
          },
          {
            type: "element",
            tagName: "span",
            properties: {
              className: ["token", "punctuation"],
            },
            children: [
              {
                type: "text",
                value: ";",
              },
            ],
          },
        ],
      };

      expect(getLines(root)).toMatchSnapshot();
    });
  });

  describe("unifyLine", () => {
    test("text, element가 섞여 있을 경우 통일되어야 함", () => {
      const line: ICodeLine = {
        type: "CODE_LINE",
        children: [
          {
            type: "TEXT",
            text: "hi",
          },
          {
            type: "CODE_ELEMENT",
            tagName: "span",
            children: [{ type: "TEXT", text: "my name is" }],
          },
          {
            type: "TEXT",
            text: "taehoon",
          },
        ],
      };

      const result = unifyLine(line);
      expect(result).toMatchSnapshot();
    });
  });
});
