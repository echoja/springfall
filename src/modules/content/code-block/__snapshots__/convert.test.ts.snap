// Vitest Snapshot v1

exports[`convert > getElement > \\n이 여러 개 있을 때 1`] = `
{
  "left": {
    "children": [
      {
        "type": "text",
        "value": "abc",
      },
    ],
    "tagName": "span1",
    "type": "element",
  },
  "reachedNewline": true,
  "right": {
    "children": [
      {
        "type": "text",
        "value": "def",
      },
      {
        "children": [
          {
            "type": "text",
            "value": "ghi
jkl",
          },
        ],
        "tagName": "span2",
        "type": "element",
      },
    ],
    "tagName": "span1",
    "type": "element",
  },
}
`;

exports[`convert > getElement > children 내에 
이 없을 때 1`] = `
{
  "node": {
    "children": [
      {
        "type": "text",
        "value": "abc",
      },
      {
        "type": "text",
        "value": "dev",
      },
    ],
    "tagName": "span",
    "type": "element",
  },
  "reachedNewline": false,
}
`;

exports[`convert > getElement > children 이 없을 떄 1`] = `
{
  "node": {
    "children": [],
    "tagName": "span",
    "type": "element",
  },
  "reachedNewline": false,
}
`;

exports[`convert > getElement > element 안에 \\n 이 포함된 Text 가 있을 떄 제대로 동작해야 함 1`] = `
{
  "left": {
    "children": [
      {
        "type": "text",
        "value": "abc",
      },
      {
        "type": "text",
        "value": "def",
      },
    ],
    "tagName": "span",
    "type": "element",
  },
  "reachedNewline": true,
  "right": {
    "children": [
      {
        "type": "text",
        "value": "ghi",
      },
      {
        "type": "text",
        "value": "jkl",
      },
    ],
    "tagName": "span",
    "type": "element",
  },
}
`;

exports[`convert > getElement > 중첩되었을 떄 마지막에 \\n이 있을 때 1`] = `
{
  "left": {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "abcdef",
              },
            ],
            "tagName": "span3",
            "type": "element",
          },
        ],
        "tagName": "span2",
        "type": "element",
      },
    ],
    "tagName": "span1",
    "type": "element",
  },
  "reachedNewline": true,
  "right": null,
}
`;

exports[`convert > getElement > 중첩된 경우 1`] = `
{
  "left": {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "abc",
              },
            ],
            "tagName": "span3",
            "type": "element",
          },
        ],
        "tagName": "span2",
        "type": "element",
      },
    ],
    "tagName": "span1",
    "type": "element",
  },
  "reachedNewline": true,
  "right": {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "def",
              },
            ],
            "tagName": "span3",
            "type": "element",
          },
        ],
        "tagName": "span2",
        "type": "element",
      },
    ],
    "tagName": "span1",
    "type": "element",
  },
}
`;

exports[`convert > getLines > 제대로 동작해야 함 1`] = `
[
  [
    {
      "children": [
        {
          "type": "text",
          "value": "let",
        },
      ],
      "properties": {
        "className": [
          "token",
          "keyword",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    {
      "type": "text",
      "value": " a ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "=",
        },
      ],
      "properties": {
        "className": [
          "token",
          "operator",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    {
      "type": "text",
      "value": " ",
    },
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "\`",
            },
          ],
          "properties": {
            "className": [
              "token",
              "template-punctuation",
              "string",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
      ],
      "properties": {
        "className": [
          "token",
          "template-string",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
  ],
  [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "   sdkfj",
            },
          ],
          "properties": {
            "className": [
              "token",
              "string",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
      ],
      "properties": {
        "className": [
          "token",
          "template-string",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
  ],
  [
    {
      "children": [
        {
          "children": [
            {
              "type": "text",
              "value": "\`",
            },
          ],
          "properties": {
            "className": [
              "token",
              "template-punctuation",
              "string",
            ],
          },
          "tagName": "span",
          "type": "element",
        },
      ],
      "properties": {
        "className": [
          "token",
          "template-string",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
    {
      "children": [
        {
          "type": "text",
          "value": ";",
        },
      ],
      "properties": {
        "className": [
          "token",
          "punctuation",
        ],
      },
      "tagName": "span",
      "type": "element",
    },
  ],
]
`;

exports[`convert > getText > \\n이 뒤에 있을 때 잘 분리되어야 함 1`] = `
{
  "left": {
    "type": "text",
    "value": "abcdef",
  },
  "reachedNewline": true,
  "right": null,
}
`;

exports[`convert > getText > \\n이 앞에 있을 때 잘 분리되어야 함 1`] = `
{
  "left": null,
  "reachedNewline": true,
  "right": {
    "type": "text",
    "value": "def",
  },
}
`;

exports[`convert > getText > \\n이 없을 때는 그대로 리턴되어야 함 1`] = `
{
  "node": {
    "type": "text",
    "value": "abcdef",
  },
  "reachedNewline": false,
}
`;

exports[`convert > getText > \\n이 있을 때 잘 분리되어야 함 1`] = `
{
  "left": {
    "type": "text",
    "value": "abc",
  },
  "reachedNewline": true,
  "right": {
    "type": "text",
    "value": "def",
  },
}
`;

exports[`convert > unifyLine > text, element가 섞여 있을 경우 통일되어야 함 1`] = `
{
  "children": [
    {
      "children": [
        {
          "text": "hi",
          "type": "TEXT",
        },
      ],
      "properties": {
        "className": [
          "token text-wrapper",
        ],
      },
      "tagName": "span",
      "type": "CODE_ELEMENT",
    },
    {
      "children": [
        {
          "text": "my name is",
          "type": "TEXT",
        },
      ],
      "tagName": "span",
      "type": "CODE_ELEMENT",
    },
    {
      "children": [
        {
          "text": "taehoon",
          "type": "TEXT",
        },
      ],
      "properties": {
        "className": [
          "token text-wrapper",
        ],
      },
      "tagName": "span",
      "type": "CODE_ELEMENT",
    },
  ],
  "type": "CODE_LINE",
}
`;
